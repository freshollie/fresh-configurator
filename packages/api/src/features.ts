import semver from "semver";
import {
  FeatureBits,
  Features,
  DisarmFlags,
  Sensors,
  SerialPortFunctions,
  EscProtocols,
  McuTypes,
  Beepers,
} from "./types";

const BASE_FEATURE_BITS: FeatureBits = {
  0: Features.RX_PPM,
  2: Features.INFLIGHT_ACC_CAL,
  3: Features.RX_SERIAL,
  4: Features.MOTOR_STOP,
  5: Features.SERVO_TILT,
  6: Features.SOFTSERIAL,
  7: Features.GPS,
  9: Features.SONAR,
  10: Features.TELEMETRY,
  12: Features["3D"],
  13: Features.RX_PARALLEL_PWM,
  14: Features.RX_MSP,
  15: Features.RSSI_ADC,
  16: Features.LED_STRIP,
  17: Features.DISPLAY,
};

export const getFeatureBits = (apiVersion: string): FeatureBits => {
  const featureBits = { ...BASE_FEATURE_BITS };
  if (!semver.gte(apiVersion, "1.33.0")) {
    featureBits[19] = Features.BLACKBOX;
  }

  if (semver.gte(apiVersion, "1.12.0")) {
    featureBits[20] = Features.CHANNEL_FORWARDING;
  }

  if (semver.gte(apiVersion, "1.15.0") && !semver.gte(apiVersion, "1.36.0")) {
    featureBits[8] = Features.FAILSAFE;
  }

  if (semver.gte(apiVersion, "1.16.0")) {
    featureBits[21] = Features.TRANSPONDER;
  }

  if (semver.gte(apiVersion, "1.16.0")) {
    featureBits[22] = Features.AIRMODE;
  }

  if (semver.gte(apiVersion, "1.16.0")) {
    if (semver.lt(apiVersion, "1.20.0")) {
      featureBits[23] = Features.SUPEREXPO_RATES;
    } else if (!semver.gte(apiVersion, "1.33.0")) {
      featureBits[23] = Features.SDCARD;
    }
  }

  if (semver.gte(apiVersion, "1.20.0")) {
    featureBits[18] = Features.OSD;
    if (!semver.gte(apiVersion, "1.35.0")) {
      featureBits[24] = Features.VTX;
    }
  }

  if (semver.gte(apiVersion, "1.31.0")) {
    featureBits[25] = Features.RX_SPI;
    featureBits[27] = Features.ESC_SENSOR;
  }

  if (semver.gte(apiVersion, "1.36.0")) {
    featureBits[28] = Features.ANTI_GRAVITY;
    featureBits[29] = Features.DYNAMIC_FILTER;
  }

  if (!semver.gte(apiVersion, "1.36.0")) {
    featureBits[1] = Features.VBAT;
    featureBits[11] = Features.VCURRENT_METER;
  }

  return featureBits;
};

export const disarmFlagBits = (apiVersion: string): DisarmFlags[] => [
  DisarmFlags.NO_GYRO,
  DisarmFlags.FAILSAFE,
  DisarmFlags.RX_FAILSAFE,
  DisarmFlags.BAD_RX_RECOVERY,
  DisarmFlags.BOXFAILSAFE,
  DisarmFlags.THROTTLE,
  ...(semver.gte(apiVersion, "1.38.0") ? [DisarmFlags.RUNAWAY_TAKEOFF] : []),

  ...(semver.gte(apiVersion, "1.42.0") ? [DisarmFlags.CRASH] : []),
  DisarmFlags.ANGLE,
  DisarmFlags.BOOT_GRACE_TIME,
  DisarmFlags.NOPREARM,
  DisarmFlags.LOAD,
  DisarmFlags.CALIBRATING,
  DisarmFlags.CLI,
  DisarmFlags.CMS_MENU,
  ...(semver.lt(apiVersion, "1.41.0") ? [DisarmFlags.OSD_MENU] : []),
  DisarmFlags.BST,
  DisarmFlags.MSP,
  ...(semver.gte(apiVersion, "1.39.0")
    ? [DisarmFlags.PARALYZE, DisarmFlags.GPS]
    : []),
  ...(semver.gte(apiVersion, "1.41.0")
    ? [DisarmFlags.RESC, DisarmFlags.RPMFILTER]
    : []),
  ...(semver.gte(apiVersion, "1.42.0")
    ? [DisarmFlags.REBOOT_REQD, DisarmFlags.DSHOT_BBANG]
    : []),
  ...(semver.gte(apiVersion, "1.43.0") ? [DisarmFlags.NO_ACC_CAL] : []),
  DisarmFlags.ARM_SWITCH,
];

export const sensorBits = (): Sensors[] => [
  Sensors.ACCELEROMETER,
  Sensors.BAROMETER,
  Sensors.MAGNETOMETER,
  Sensors.GPS,
  Sensors.SONAR,
  Sensors.GYRO,
];

export const serialPortFunctionBits = (): SerialPortFunctions[] => [
  SerialPortFunctions.MSP,
  SerialPortFunctions.GPS,
  SerialPortFunctions.TELEMETRY_FRSKY,
  SerialPortFunctions.TELEMETRY_HOTT,
  SerialPortFunctions.TELEMETRY_LTM,
  SerialPortFunctions.TELEMETRY_SMARTPORT,
  SerialPortFunctions.RX_SERIAL,
  SerialPortFunctions.BLACKBOX,
  SerialPortFunctions.UNKNOWN,
  SerialPortFunctions.TELEMETRY_MAVLINK,
  SerialPortFunctions.ESC_SENSOR,
  SerialPortFunctions.TBS_SMARTAUDIO,
  SerialPortFunctions.TELEMETRY_IBUS,
  SerialPortFunctions.IRC_TRAMP,
  SerialPortFunctions.RUNCAM_DEVICE_CONTROL,
  SerialPortFunctions.LIDAR_TF,
];

export const beeperBits = (apiVersion: string): Beepers[] => [
  Beepers.GYRO_CALIBRATED,
  Beepers.RX_LOST,
  Beepers.RX_LOST_LANDING,
  Beepers.DISARMING,
  Beepers.ARMING,
  Beepers.ARMING_GPS_FIX,
  Beepers.BAT_CRIT_LOW,
  Beepers.BAT_LOW,
  Beepers.GPS_STATUS,
  Beepers.RX_SET,
  Beepers.ACC_CALIBRATION,
  Beepers.ACC_CALIBRATION_FAIL,
  Beepers.READY_BEEP,
  Beepers.MULTI_BEEPS,
  Beepers.DISARM_REPEAT,
  Beepers.ARMED,
  Beepers.SYSTEM_INIT,
  Beepers.USB,
  Beepers.BLACKBOX_ERASE,
  ...(semver.gte(apiVersion, "1.37.0")
    ? [
        Beepers.CRASH_FLIP,
        Beepers.CAM_CONNECTION_OPEN,
        Beepers.CAM_CONNECTION_CLOSE,
      ]
    : []),
  ...(semver.gte(apiVersion, "1.39.0") ? [Beepers.RC_SMOOTHING_INIT_FAIL] : []),
];

export const legacySerialPortFunctionsMap: Record<
  number,
  SerialPortFunctions[] | undefined
> = {
  1: [SerialPortFunctions.MSP],
  5: [SerialPortFunctions.MSP],
  8: [SerialPortFunctions.MSP],
  2: [SerialPortFunctions.GPS],
  3: [SerialPortFunctions.RX_SERIAL],
  10: [SerialPortFunctions.BLACKBOX],
  11: [SerialPortFunctions.MSP, SerialPortFunctions.BLACKBOX],
};

export const escProtocols = (version: string): EscProtocols[] => [
  EscProtocols.PWM,
  EscProtocols.ONESHOT125,
  EscProtocols.ONESHOT42,
  EscProtocols.MULTISHOT,
  ...(semver.gte(version, "1.20.0") ? [EscProtocols.BRUSHED] : []),
  ...(semver.gte(version, "1.31.0")
    ? [EscProtocols.DSHOT150, EscProtocols.DSHOT300, EscProtocols.DSHOT600]
    : []),
  ...(semver.gte(version, "1.31.0") && semver.lt(version, "1.42.0")
    ? [EscProtocols.DSHOT1200]
    : []),
  ...(semver.gte(version, "1.36.0") ? [EscProtocols.PROSHOT1000] : []),
  ...(semver.gte(version, "1.43.0") ? [EscProtocols.DISABLED] : []),
];

export const MCU_GROUPS = {
  F7: [McuTypes.F722, McuTypes.F745, McuTypes.F746, McuTypes.F765],
  F3: [McuTypes.F303],
  F4: [McuTypes.F411, McuTypes.F446],
  H7: [
    McuTypes.H723_725,
    McuTypes.H743_REV_UNKNOWN,
    McuTypes.H743_REV_V,
    McuTypes.H743_REV_X,
    McuTypes.H743_REV_Y,
    McuTypes.H750,
    McuTypes.H7A3,
  ],
  F1: [McuTypes.F103],
};

export const mcuGroupFromId = (
  mcuTypeId: McuTypes
): keyof typeof MCU_GROUPS | undefined =>
  (Object.keys(MCU_GROUPS) as (keyof typeof MCU_GROUPS)[]).find((key) =>
    MCU_GROUPS[key].includes(mcuTypeId)
  );

export const MIXER_LIST = [
  { name: "Tricopter", id: 3, model: "tricopter", image: "tri" },
  { name: "Quad +", id: 2, model: "quad_x", image: "quad_p" },
  { name: "Quad X", id: 0, model: "quad_x", image: "quad_x" },
  { name: "Bicopter", id: 16, model: "custom", image: "bicopter" },
  { name: "Gimbal", id: 4, model: "custom", image: "custom" },
  { name: "Y6", id: 20, model: "y6", image: "y6" },
  { name: "Hex +", id: 5, model: "hex_plus", image: "hex_p" },
  { name: "Flying Wing", id: 10, model: "custom", image: "flying_wing" },
  { name: "Y4", id: 19, model: "y4", image: "y4" },
  { name: "Hex X", id: 6, model: "hex_x", image: "hex_x" },
  { name: "Octo X8", id: 21, model: "custom", image: "octo_x8" },
  { name: "Octo Flat +", id: 8, model: "custom", image: "octo_flat_p" },
  { name: "Octo Flat X", id: 9, model: "custom", image: "octo_flat_x" },
  { name: "Airplane", id: 11, model: "custom", image: "airplane" },
  { name: "Heli 120", id: 12, model: "custom", image: "custom" },
  { name: "Heli 90", id: 13, model: "custom", image: "custom" },
  { name: "V-tail Quad", id: 17, model: "quad_vtail", image: "vtail_quad" },
  { name: "Hex H", id: 7, model: "custom", image: "custom" },
  { name: "PPM to SERVO", id: 22, model: "custom", image: "custom" },
  { name: "Dualcopter", id: 15, model: "custom", image: "custom" },
  { name: "Singlecopter", id: 14, model: "custom", image: "custom" },
  { name: "A-tail Quad", id: 18, model: "quad_atail", image: "atail_quad" },
  { name: "Custom", id: 23, model: "custom", image: "custom" },
  { name: "Custom Airplane", id: 24, model: "custom", image: "custom" },
  { name: "Custom Tricopter", id: 25, model: "custom", image: "custom" },
  { name: "Quad X 1234", id: 1, model: "quad_x", image: "quad_x_1234" },
];
