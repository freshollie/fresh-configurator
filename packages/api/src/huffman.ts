const HUFFMAN_EOF = -1;
const tree: Record<number, Record<number, number>> = {
  "2": { 0b11: 0 },
  "3": { 0b101: 1 },
  "4": { 0b1001: 2 },
  "5": { 0b10001: 3, 0b10000: 4, 0b01111: 80 },
  "6": {
    0b011101: 5,
    0b011100: 6,
    0b011011: 7,
    0b011010: 8,
    0b011001: 16,
  },
  "7": {
    0b0110001: 9,
    0b0110000: 10,
    0b0101111: 11,
    0b0101110: 12,
    0b0101101: 13,
    0b0101100: 14,
    0b0101011: 15,
    0b0101010: 17,
    0b0101001: 18,
  },
  "8": {
    0b01010001: 19,
    0b01010000: 20,
    0b01001111: 21,
    0b01001110: 22,
    0b01001101: 23,
    0b01001100: 24,
    0b01001011: 25,
    0b01001010: 26,
    0b01001001: 27,
    0b01001000: 28,
    0b01000111: 29,
    0b01000110: 30,
    0b01000101: 31,
    0b01000100: 32,
    0b01000011: 33,
    0b01000010: 34,
    0b01000001: 35,
    0b01000000: 36,
    0b00111111: 48,
    0b00111110: 64,
    0b00111101: 240,
  },
  "9": {
    0b001111001: 37,
    0b001111000: 38,
    0b001110111: 39,
    0b001110110: 40,
    0b001110101: 41,
    0b001110100: 42,
    0b001110011: 43,
    0b001110010: 44,
    0b001110001: 45,
    0b001110000: 46,
    0b001101111: 47,
    0b001101110: 49,
    0b001101101: 50,
    0b001101100: 51,
    0b001101011: 52,
    0b001101010: 53,
    0b001101001: 54,
    0b001101000: 55,
    0b001100111: 56,
    0b001100110: 57,
    0b001100101: 58,
    0b001100100: 59,
    0b001100011: 60,
    0b001100010: 61,
    0b001100001: 62,
    0b001100000: 63,
    0b001011111: 65,
    0b001011110: 66,
    0b001011101: 67,
    0b001011100: 68,
    0b001011011: 69,
    0b001011010: 70,
    0b001011001: 71,
    0b001011000: 72,
    0b001010111: 73,
    0b001010110: 76,
    0b001010101: 79,
    0b001010100: 81,
    0b001010011: 128,
    0b001010010: 224,
    0b001010001: 241,
    0b001010000: 255,
  },
  "10": {
    0b0010011111: 74,
    0b0010011110: 75,
    0b0010011101: 77,
    0b0010011100: 78,
    0b0010011011: 82,
    0b0010011010: 83,
    0b0010011001: 84,
    0b0010011000: 85,
    0b0010010111: 86,
    0b0010010110: 87,
    0b0010010101: 88,
    0b0010010100: 89,
    0b0010010011: 90,
    0b0010010010: 91,
    0b0010010001: 92,
    0b0010010000: 93,
    0b0010001111: 94,
    0b0010001110: 95,
    0b0010001101: 96,
    0b0010001100: 97,
    0b0010001011: 98,
    0b0010001010: 99,
    0b0010001001: 100,
    0b0010001000: 101,
    0b0010000111: 102,
    0b0010000110: 103,
    0b0010000101: 104,
    0b0010000100: 105,
    0b0010000011: 106,
    0b0010000010: 107,
    0b0010000001: 108,
    0b0010000000: 109,
    0b0001111111: 110,
    0b0001111110: 111,
    0b0001111101: 112,
    0b0001111100: 113,
    0b0001111011: 114,
    0b0001111010: 115,
    0b0001111001: 116,
    0b0001111000: 117,
    0b0001110111: 118,
    0b0001110110: 119,
    0b0001110101: 120,
    0b0001110100: 121,
    0b0001110011: 122,
    0b0001110010: 123,
    0b0001110001: 124,
    0b0001110000: 125,
    0b0001101111: 126,
    0b0001101110: 127,
    0b0001101101: 129,
    0b0001101100: 130,
    0b0001101011: 131,
    0b0001101010: 132,
    0b0001101001: 133,
    0b0001101000: 134,
    0b0001100111: 135,
    0b0001100110: 136,
    0b0001100101: 137,
    0b0001100100: 138,
    0b0001100011: 139,
    0b0001100010: 140,
    0b0001100001: 141,
    0b0001100000: 142,
    0b0001011111: 143,
    0b0001011110: 144,
    0b0001011101: 145,
    0b0001011100: 146,
    0b0001011011: 147,
    0b0001011010: 148,
    0b0001011001: 149,
    0b0001011000: 150,
    0b0001010111: 151,
    0b0001010110: 152,
    0b0001010101: 153,
    0b0001010100: 154,
    0b0001010011: 155,
    0b0001010010: 156,
    0b0001010001: 157,
    0b0001010000: 158,
    0b0001001111: 159,
    0b0001001110: 160,
    0b0001001101: 161,
    0b0001001100: 162,
    0b0001001011: 163,
    0b0001001010: 164,
    0b0001001001: 165,
    0b0001001000: 166,
    0b0001000111: 167,
    0b0001000110: 168,
    0b0001000101: 169,
    0b0001000100: 170,
    0b0001000011: 171,
    0b0001000010: 172,
    0b0001000001: 173,
    0b0001000000: 174,
    0b0000111111: 175,
    0b0000111110: 176,
    0b0000111101: 177,
    0b0000111100: 178,
    0b0000111011: 179,
    0b0000111010: 180,
    0b0000111001: 181,
    0b0000111000: 182,
    0b0000110111: 183,
    0b0000110110: 184,
    0b0000110101: 185,
    0b0000110100: 186,
    0b0000110011: 187,
    0b0000110010: 188,
    0b0000110001: 189,
    0b0000110000: 190,
    0b0000101111: 191,
    0b0000101110: 192,
    0b0000101101: 193,
    0b0000101100: 194,
    0b0000101011: 195,
    0b0000101010: 196,
    0b0000101001: 197,
    0b0000101000: 198,
    0b0000100111: 199,
    0b0000100110: 200,
    0b0000100101: 201,
    0b0000100100: 202,
    0b0000100011: 203,
    0b0000100010: 204,
    0b0000100001: 205,
    0b0000100000: 206,
    0b0000011111: 207,
    0b0000011110: 208,
    0b0000011101: 209,
    0b0000011100: 210,
    0b0000011011: 211,
    0b0000011010: 212,
    0b0000011001: 214,
    0b0000011000: 215,
    0b0000010111: 216,
    0b0000010110: 217,
    0b0000010101: 218,
    0b0000010100: 219,
    0b0000010011: 220,
    0b0000010010: 222,
    0b0000010001: 223,
    0b0000010000: 225,
    0b0000001111: 226,
    0b0000001110: 228,
    0b0000001101: 239,
  },
  "11": {
    0b00000011001: 213,
    0b00000011000: 221,
    0b00000010111: 227,
    0b00000010110: 229,
    0b00000010101: 230,
    0b00000010100: 231,
    0b00000010011: 232,
    0b00000010010: 233,
    0b00000010001: 234,
    0b00000010000: 235,
    0b00000001111: 236,
    0b00000001110: 237,
    0b00000001101: 238,
    0b00000001100: 242,
    0b00000001011: 243,
    0b00000001010: 244,
    0b00000001001: 245,
    0b00000001000: 246,
    0b00000000111: 247,
    0b00000000110: 248,
    0b00000000101: 250,
    0b00000000100: 251,
    0b00000000011: 252,
    0b00000000010: 253,
    0b00000000001: 254,
  },
  "12": { 0b000000000001: 249, 0b000000000000: HUFFMAN_EOF },
};

// eslint-disable-next-line import/prefer-default-export
export const huffmanDecodeBuffer = (
  inBuf: Buffer,
  expectedSize: number
): Buffer => {
  let code = 0;
  let depth = 0;
  let bitOffset = 0;
  let currentByte = 0;
  const outBuf = [];

  while (bitOffset / 8 < inBuf.byteLength) {
    if (outBuf.length === expectedSize) {
      // we've exhausted the input stream, discard any odd bits on the end
      break;
    }

    if (bitOffset % 8 === 0) {
      // move ono the next byte
      currentByte = inBuf.readInt8(bitOffset / 8);
    }

    // get the next bit from the input buffer
    code <<= 1;
    depth += 1;
    // if the next bit is one, set to one 000 -> 001
    if ((128 >> bitOffset % 8) & currentByte) {
      code |= 0x01;
    }

    const decodedValue = tree[depth]?.[code];
    if (decodedValue !== undefined) {
      if (decodedValue === HUFFMAN_EOF) {
        break;
      } else {
        // output the value
        outBuf.push(decodedValue);
      }

      // reset the code to continue decompressing the input buffer
      code = 0;
      depth = 0;
    }
    bitOffset += 1;
  }

  return Buffer.from(outBuf);
};
