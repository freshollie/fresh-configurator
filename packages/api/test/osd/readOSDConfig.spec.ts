import mockMsp from "../mockMsp";
import readOsdData from "./osdConfigData";
import { readOSDConfig } from "../../src";
import codes from "../../src/codes";

describe("readOSDConfig", () => {
  it("should read a 1.42.0 config correctly", async () => {
    mockMsp.setResponse(readOsdData);
    mockMsp.setApiVersion("1.42.0");

    const config = await readOSDConfig("/dev/something");
    expect(config).toMatchSnapshot();
    expect(mockMsp.execute).toHaveBeenCalledWith("/dev/something", {
      code: codes.MSP_OSD_CONFIG,
    });
  });
});
