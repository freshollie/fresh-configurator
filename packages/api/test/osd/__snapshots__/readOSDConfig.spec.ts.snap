// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readOSDConfig should read a 1.42.0 config correctly 1`] = `
Object {
  "alarms": Array [
    Object {
      "key": 0,
      "value": 25,
    },
    Object {
      "key": 1,
      "value": 2200,
    },
    Object {
      "key": 3,
      "value": 100,
    },
  ],
  "displayItems": Array [
    Object {
      "key": 0,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 1,
      "position": Object {
        "x": 22,
        "y": 11,
      },
      "visibilityProfiles": Array [
        true,
        false,
        false,
      ],
    },
    Object {
      "key": 2,
      "position": Object {
        "x": 13,
        "y": 6,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 3,
      "position": Object {
        "x": 14,
        "y": 2,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 4,
      "position": Object {
        "x": 14,
        "y": 6,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 18,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 19,
      "position": Object {
        "x": 1,
        "y": 11,
      },
      "visibilityProfiles": Array [
        true,
        false,
        false,
      ],
    },
    Object {
      "key": 20,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 14,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 6,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 8,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 12,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 13,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 21,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 22,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 15,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 23,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 24,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 25,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 26,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 27,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 28,
      "position": Object {
        "x": 9,
        "y": 9,
      },
      "visibilityProfiles": Array [
        true,
        true,
        true,
      ],
    },
    Object {
      "key": 30,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 31,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 32,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 33,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 34,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 35,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 36,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 11,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 37,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 38,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 39,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 40,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 41,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 42,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 43,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 44,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 45,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 46,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 47,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 48,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 49,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 50,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 51,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 52,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 53,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 54,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 55,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 56,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 57,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 58,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 59,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 60,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 61,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 62,
      "position": Object {
        "x": 10,
        "y": 7,
      },
      "visibilityProfiles": Array [
        true,
        false,
        false,
      ],
    },
  ],
  "flags": Object {
    "hasOSD": true,
    "haveMax7456Video": true,
    "haveOsdFeature": true,
    "isMax7456Detected": true,
    "isOsdSlave": false,
  },
  "osdProfiles": Object {
    "count": 3,
    "selected": 1,
  },
  "parameters": Object {
    "cameraFrameHeight": 11,
    "cameraFrameWidth": 24,
    "overlayRadioMode": 2,
  },
  "statisticItems": Array [
    Object {
      "enabled": false,
      "key": 12,
    },
    Object {
      "enabled": false,
      "key": 8,
    },
    Object {
      "enabled": true,
      "key": 9,
    },
    Object {
      "enabled": true,
      "key": 0,
    },
    Object {
      "enabled": false,
      "key": 10,
    },
    Object {
      "enabled": true,
      "key": 1,
    },
    Object {
      "enabled": false,
      "key": 7,
    },
    Object {
      "enabled": false,
      "key": 13,
    },
    Object {
      "enabled": true,
      "key": 2,
    },
    Object {
      "enabled": true,
      "key": 3,
    },
    Object {
      "enabled": true,
      "key": 4,
    },
    Object {
      "enabled": false,
      "key": 5,
    },
    Object {
      "enabled": true,
      "key": 6,
    },
    Object {
      "enabled": true,
      "key": 11,
    },
    Object {
      "enabled": false,
      "key": 14,
    },
    Object {
      "enabled": false,
      "key": 15,
    },
    Object {
      "enabled": false,
      "key": 16,
    },
    Object {
      "enabled": false,
      "key": 17,
    },
    Object {
      "enabled": false,
      "key": 18,
    },
    Object {
      "enabled": false,
      "key": 19,
    },
    Object {
      "enabled": false,
      "key": 20,
    },
    Object {
      "enabled": false,
      "key": 21,
    },
    Object {
      "enabled": false,
      "key": 22,
    },
    Object {
      "enabled": false,
      "key": 23,
    },
  ],
  "timers": Array [
    Object {
      "key": 0,
      "precision": 0,
      "src": 0,
      "time": 10,
    },
    Object {
      "key": 1,
      "precision": 0,
      "src": 1,
      "time": 10,
    },
  ],
  "unitMode": 1,
  "videoSystem": 2,
  "warnings": Array [
    Object {
      "enabled": true,
      "key": 0,
    },
    Object {
      "enabled": true,
      "key": 1,
    },
    Object {
      "enabled": true,
      "key": 2,
    },
    Object {
      "enabled": true,
      "key": 3,
    },
    Object {
      "enabled": true,
      "key": 4,
    },
    Object {
      "enabled": true,
      "key": 5,
    },
    Object {
      "enabled": true,
      "key": 6,
    },
    Object {
      "enabled": true,
      "key": 7,
    },
    Object {
      "enabled": true,
      "key": 8,
    },
    Object {
      "enabled": true,
      "key": 9,
    },
    Object {
      "enabled": true,
      "key": 10,
    },
    Object {
      "enabled": true,
      "key": 11,
    },
    Object {
      "enabled": true,
      "key": 12,
    },
    Object {
      "enabled": false,
      "key": 13,
    },
    Object {
      "enabled": false,
      "key": 14,
    },
    Object {
      "enabled": false,
      "key": 15,
    },
  ],
}
`;
